extract machine event from database backups

zhenz opened the pull request.

zhenz committed:
extract machine event from database backups

diurnalist commented on starcompactor/api_dump.py:
```diff
@@ -13,23 +13,26 @@
     # parser.add_argument('-s', '--start', type=str)
     # parser.add_argument('-e', '--end', type=str)
-    parser.add_argument('-c', '--epoch', type=str, default='2015-09-06T23:31:16',
+    parser.add_argument('-c', '--epoch', type=str, default=Constants.CHAMELEON_KVM_START_DATE + 'T00:00:00',
```
+1 for pulling out as a constant

diurnalist commented on starcompactor/extractors/machine.py:
```diff
@@ -0,0 +1,152 @@
+BACKUP_FILE_PREFIX = 'mysql_backup_{date}'
+NOVA_COMPUTE_HOST_BINARY = 'nova-compute'
+
+def get_machine_event(process_no, start_date, end_date, backup_path):
```
What is the reason for parsing the text-formatted backup file, rather than reading this from an actual database? Parsing the text file will break if any columns are reordered, or if a new column is added as the first column.

Also, where are these backups generated? Does the script do that? Or are these ones that we had previously made? It looks like they are assumed to be in a very specific format, which is not obvious and would be difficult for others to run.

An alternate solution which gets around these constraints is to have the user provide the name of a database to read from. That way the query operation is much simpler and less error-prone.

zhenz replied:
@diurnalist I tried importing to an actual database from backup sql file, but it would be very slow. Processing one file would take several minutes. Considering we have 4 years of backup files, it would take an extremely long time. I tried to run multiple files in parallel but didn't help. Parsing sql file directly speeds things up a lot. 

When I parse the sql file, I record the column order, so if columns are reordered or new columns are added, it would capture that. 

The backup files are our daily backups on UC storage node.

zhenz replied:
@diurnalist reading text-formatted file is slow as well. I guess the data grows a lot compared to the last time I ran the script. I'll try to optimize it.

diurnalist replied:
I suspect it might be slow because it's trying to import the entire database, hm. I wonder if it's possible to tell mysql to only import a few tables from a dump. Something like [this approach](https://tecadmin.net/extract-tables-backup-from-full-mysql-dump/) might work. It involves first splitting the dump coarsely into per-table dumps. That allows you to then import only the tables you need. I think the splitting of the file can probably be done quite quickly, but we'd need to try it out.

diurnalist commented on starcompactor/extractors/machine.py:
```diff
@@ -0,0 +1,152 @@
+        
+        if len(backup_files) == 0: LOG.warning('Missing date {}'.format(filedate))
+        
+        for f in backup_files:
```
Do newer backup files have a copy of data existing in the older files? That's typically the case with a full snapshot. If so, then you'll be processing the same events multiple times. I see there is a map-like object being used to store the events, which helps somewhat. It still feels better to only look at events once--is that possible?

zhenz commented:
@diurnalist I fixed the script by importing the database and run queries to get info, instead of reading the plain text of SQL. Also, I managed to only import the tables I need to optimize the running time. I updated the README as well. Could you review it again? Thank you!

diurnalist commented on starcompactor/extractors/machine.py:
```diff
@@ -0,0 +1,101 @@
+    tmp_sql_file_name = 'kvm-{}.sql'.format(os.path.basename(backup_file).split('.')[0])
+    
+    try:
+        with open_by_suffix(backup_file) as f_in:
```
If you just need to copy the file to a temp location, perhaps [`shutil.copy`](https://docs.python.org/3/library/shutil.html#shutil.copy) is the better tool?

diurnalist commented on starcompactor/extractors/machine.py:
```diff
@@ -0,0 +1,101 @@
+    try:
+        db.cursor.execute(extract_data_sql)
+        for row in db.cursor:
+            if not row[5] or row[5] != NOVA_COMPUTE_HOST_BINARY:
```
I think it will be easier to avoid mistakes in the future if we pull these out to named variables. Something like:

```python
created_at, node_updated_at, deleted_at, service_updated_at, disabled, binary, hypervisor_hostname, vcpus, memory_mb, local_gb = row
```

Maybe there is a more eloquent way. Can also just do like

```python
created_at = row[0]
node_updated_at = row[1]
# ... etc
```

diurnalist commented on starcompactor/machine_event_dump.py:
```diff
@@ -0,0 +1,204 @@
+        machine_args.append([process_no, backup_file, mysqlargs.connect_kwargs, rack_extractor])  
+        process_no = process_no + 1
+    
+    with contextlib.closing(multiprocessing.Pool(processes=int(math.ceil(process_no / int(config.get('multithread', 'number_of_files_per_process')))))) as pool:
```
Nice use of multiprocessing 👍

diurnalist commented on starcompactor/machine_event_dump.py:
```diff
@@ -0,0 +1,204 @@
+    # Example:
+    # In 2015-10-09 backup and for host x, "created_at" 2015-10-09 with 48 available VCPUs.
+    # In 2017-11-01 backup and for host x, "created_at" 2015-10-09 with 40 available VCPUs.
+    # For create event of host x, we choose the content from 2015-10-09 backup.
```
Good comment!

diurnalist commented on starcompactor/machine_event_dump.py:
```diff
@@ -0,0 +1,204 @@
+    # For timestamps we extracted from the backups, they might not be the timestamps when 
+    # updates of the fields that we concerned occurred. 
+    # Therefore, we need to refine the records we collected from the backup files, and use the
+    # the most possible timestamps as the update timestamps of the fields that we concerned.  
```
earliest possible?

diurnalist commented on starcompactor/machine_event_dump.py:
```diff
@@ -0,0 +1,204 @@
+
+def refine_event(machine_events, event1, event2):
+    event_by_host = {}
+    for key in machine_events.keys():
```
Took me a long while to figure out what this is doing. I think this code could benefit from the `machine_events` being preprocessed to be dicts rather than arrays. What are the `values()` of machine_events?

Also, I'm not sure if this function really does anything in the case where `event2` is None? I think it might not need to run at all in this case. From what I gather, it loops over all the events and categorizes them per host, and stores the value for the time as a tuple, which will be unique for each event, which means the following loop will always execute and add the tuple _back_ into the update dict, effectively making it update itself with the same contents? Unless I'm reading that wrong.

With all that said, this is how I understood the function, is it accurate?

```python
# Filter by events in the input couple
for evt in machine_events if evt['EVENT'] in [event1, event2]:
  hostname = event['HOSTNAME']
  time = event['EVENT_TIME']
  # Create entries by host, by time, for the event (type) that occurred at that time
  host_events = event_by_host.setdefault(hostname, {})
  host_events[time] = evt['EVENT']

# Now, take those grouped events to figure out which events
# to keep (we are throwing away intermediate events that didn't
# change anything or appear erroneous
for host, host_events in event_by_host.items():
  current_content = None
  for event_time in sorted(host_events):
    content = host_events[event_time]
    if content != current_content:
      keep_event_keys.append((event_time, host, content))
    current_content = content

return { key: machine_events[key] for key in keep_event_keys }
```

Also also, for the second loop, in case of some string like "CREATE", "CREATE", "DELETE", it will take the first CREATE. Should it? Is the second create more valid? (What if the first CREATE resulted in some application error?)

diurnalist commented on starcompactor/machine_event_dump.py:
```diff
@@ -0,0 +1,204 @@
+    # Therefore, we need to refine the records we collected from the backup files, and use the
+    # the most possible timestamps as the update timestamps of the fields that we concerned.  
+    refined_machine_events = {}
+    refined_machine_events.update(refine_event(machine_events, 'CREATE', 'DELETE'))
```
I think I finally get what this does now! It may be helpful to attack this problem with a smarter data structure so you don't have to loop over all the machine events 3 times. Not sure how big these lists are, so unsure how much time savings you'll get. Here's what I think is equivalent code to what you have, after thinking about it (sort of pseudocode, have not tested.) I think it manages to run through all the events once (+ a sort operation). This should do the job of both `refine_event` and `refine_for_initial_create`.

```python
"""
Keep track of all events and the current "state" of an open/close pair, e.g. CREATE/DELETE.
When a pair is opened, the state is updated to indicate this. Turns out we only need
to keep track of the state we came from, because there is only 1 possibility there. We can
also keep track of the last "measures" we got, so we can throw away future events with
unchanged data (which we also want to do?)
{ 
  <host>: {
    'state': {
      'CREATE': { waiting: <bool>, prev: 'DELETE' },
      'ENABLE': { waiting: <bool>, prev: 'DISABLE' },
      'DELETE': { waiting: <bool>, prev: 'CREATE' },
      'DISABLE': { waiting: <bool>, prev: 'ENABLE' },
    },
    'measures': <event>
  }
}
"""
class HostProcessor():
  def __init__(self):
    self.states = {
      'CREATE': { 'waiting': False, 'prev': 'DELETE' },
      'ENABLE': { 'waiting': False, 'prev': 'DISABLE' },
      'DELETE': { 'waiting': False, 'prev': 'CREATE' },
      'DISABLE': { 'waiting': False, 'prev': 'ENABLE' },
    }
    self.measures = None

  def process(self, event_type, measures):
    # Check for illegal events due to state
    if event_type in self.state:
      if self.state[event_type]['waiting']:
        return False
      # Toggle waiting state between opposing pair
      self.state[state[event_type]['prev']]['waiting'] = False
      self.state[event_type]['waiting'] = True
    # Check for unimportant events due to no changes
    if event_type in ['CREATE', 'UPDATE', 'ENABLE']:
      if self.measures == measures:
        return False
      self.measures = measures
    return True
    
# Sort events by time (could be done in MySQL as well)
def by_time(item):
  # Sort by 2nd element of dict key tuple
  return item[0][1]

host_processors = {}
def should_keep(host, event_type, measures):
  processor = host_processors.setdefault(host, new HostProcessor)
  return processor.process(event_type, measures)

refined_events = { 
  key: measures 
  for key, measures in sorted(machine_events, key=by_time).items() 
  if should_keep(key[1], key[2], measures)
}
```

diurnalist reviewed (approved):
This task was quite the puzzle! Once I got what this is trying to do (and I hope I got it), I got nerd-sniped:

![nerd_sniping](https://imgs.xkcd.com/comics/nerd_sniping.png)

I left some comments on how the code could be approached differently, but correctness is more important, and it looks like what you have does the job, so great work! 🎆
